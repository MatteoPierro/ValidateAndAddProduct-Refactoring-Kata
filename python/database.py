
class DatabaseAccess:
    def store_product(self, product):
        raise Exception("This method cannot be run from a unit test since it accesses the database")
