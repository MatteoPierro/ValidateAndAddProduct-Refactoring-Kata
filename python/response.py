
class ProductFormData:

    def __init__(self, name, product_type, weight, suggested_price, packaging_recyclable):
        self.name = name
        self.product_type = product_type
        self.weight = weight
        self.suggested_price = suggested_price
        self.packaging_recyclable = packaging_recyclable


class Response:
    """Sent to the client to indicate whether the new product was added successfully."""

    def __init__(self, product_id, status_code, message):
        self.product_id = product_id
        self.status_code = status_code
        self.message = message

    def __repr__(self) -> str:
        return f"Response("\
               f"productId={self.product_id},"\
               f" statusCode={self.status_code}"\
               f" message='{self.message}'"\
               ')'

