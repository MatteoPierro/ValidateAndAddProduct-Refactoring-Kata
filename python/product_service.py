from model_objects import Product, ProductFamily, ProductRange
from response import Response, ProductFormData


def validate_and_add(product_data: ProductFormData, db):
    if product_data.name == "":
        return Response(0, -2, "Missing Name")
    if product_data.product_type == "":
        return Response(0, -2, "Missing Type")

    product = Product(product_data.name)
    product.product_type = "Unknown"

    if "Eyeshadow" == product_data.product_type or "Mascara" == product_data.product_type:
        product.product_type = product_data.product_type
        product.product_family = ProductFamily.EYES
        if "Eyeshadow" == product_data.product_type and product.name.contains("Queen"):
            product.product_range = ProductRange.QUEEN

    product.range = ProductRange.BUDGET
    if product_data.packaging_recyclable:
        product.range = ProductRange.PROFESSIONAL

    if "Foundation" == product_data.product_type:
        if product_data.suggested_price > 10:
            product.range = ProductRange.PROFESSIONAL

    if "Lipstick" == product_data.product_type:
        product.product_type = product_data.product_type
        product.product_family = ProductFamily.LIPS

    if product_data.suggested_price > 10:
        product.range = ProductRange.PROFESSIONAL

    if product_data.suggested_price > 20:
        if 0 < product_data.weight < 10:
            return Response(0, -1, "Error - failed quality check for Queen Range")
        product.range = ProductRange.QUEEN

    if "Mascara" == product_data.product_type:
        product.product_family = ProductFamily.LASHES
        if product_data.suggested_price > 15:
            product.range = ProductRange.PROFESSIONAL

        if product_data.suggested_price > 25 and product_data.packaging_recyclable:
            product.range = ProductRange.QUEEN

    if product_data.weight < 0:
        return Response(0, -3, "Weight error")

    product.weight = product_data.weight

    if "Blusher" == product_data.product_type or "Foundation" == product_data.product_type:
        product.product_type = product_data.product_type
        product.product_family = ProductFamily.SKIN
        if "Blusher" == product_data.product_type and product_data.weight > 10:
            return Response(0, -3, "Error - weight too high")

    if not product_data.packaging_recyclable and product.range == ProductRange.QUEEN:
        return Response(0, -1, "Error - failed quality check for Queen Range")

    if "Unknown" == product.product_type:
        return Response(0, -1, f"Unknown product type {product_data.product_type}")

    return Response(db.store_product(product), 0, "Product Successfully Added")

