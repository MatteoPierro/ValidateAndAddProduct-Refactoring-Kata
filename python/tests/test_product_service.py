from approvaltests import verify_all_combinations

from database import DatabaseAccess
from product_service import validate_and_add
from response import ProductFormData


class FakeDatabase(DatabaseAccess):
    def __init__(self):
        self.product = None

    def store_product(self, product):
        self.product = product
        return 1


def do_validate_and_add(name, product_type, weight, price, recyclable):
    product_data = ProductFormData(name, product_type, weight, price, recyclable)
    db = FakeDatabase()
    response = validate_and_add(product_data, db)
    return f"{response} {db.product}"


def test_validate_and_add():
    verify_all_combinations(do_validate_and_add,
                            [["", "Sample product", "Sample Queen product"],
                             ["", "Lipstick", "Eyeshadow", "Mascara", "Blusher", "Foundation", "Unknown"],
                             [-1, 5, 11],
                             [10, 11, 21, 26],
                             [False, True]
                             ]
                            )
