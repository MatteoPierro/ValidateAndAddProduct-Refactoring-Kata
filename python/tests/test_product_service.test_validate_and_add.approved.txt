Response(productId=1, statusCode=0 message='Product Successfully Added') Product(name='Sample product', type='Lipstick', weight=5, family=ProductFamily.LIPS, range=None')